### PR TITLE
handle diverging functions forwarding their return place

### LINKED ISSUE #1075: Running diverging closure coerced to fn ptr fails: "invalid use of NULL pointer"
The following code fails to run in Miri:
```rust
fn main() {
    let f: fn() -> ! = || std::process::exit(0);  
    f();
}
```
The error is
```
error: Miri evaluation error: invalid use of NULL pointer
   --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/ops/function.rs:227:5
    |
227 |     extern "rust-call" fn call_once(self, args: Args) -> Self::Output;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Miri evaluation error: invalid use of NULL pointer
    |
note: inside call to `<[closure@src/main.rs:2:24: 2:48] as std::ops::FnOnce<()>>::call_once - shim` at src/main.rs:3:5
   --> src/main.rs:3:5
    |
3   |     f();
    |     ^^^
```
I am not entirely sure what is happening; it looks like this has something to do with the shim that is used to coerce non-capturing closures to function pointers.

Cc @eddyb @oli-obk

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/miri/issues/1075: the shim around diverging closures turned into function pointers actually "obtains" a return place inside a diverging function, but just uses it as the return place for a diverging callee. Handle this by using NULL places.

This is kind of a hack as it breaks our invariant that all places are dereferencable, but we'd eventually let raw pointers break that anyway I assume so that seems fine.

r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/interpret/place.rs [hunk 1]
```diff
@@ -651,20 +651,28 @@ where
...
+                // their layout on return.
+                PlaceTy {
+                    place: match self.frame().return_place {
+                        Some(p) => *p,
+                        // Even if we don't have a return place, we sometimes need to
+                        // create this place, but any attempt to read from / write to it
+                        // (even a ZST read/write) needs to error, so let us make this
+                        // a NULL place.
+                        //
+                        // FIXME: Ideally we'd make sure that the place projections also
```
Comment (oli-obk) [inline src/librustc_mir/interpret/place.rs]: Like just doing a place projection on a null place should error?

Comment (RalfJung) [reply src/librustc_mir/interpret/place.rs]: Place projections should have `inbounds` rules, basically. We have an [implementation of that](https://github.com/rust-lang/miri/blob/3bf51f591f7ffdd45f51cd1d42a4002482af2bd5/src/operator.rs#L103) in Miri, and these days I am actually not ashamed of it any more. ;)  

But I am worried about the perf impact of this and I think it is currently actually impossible to violate the inbounds requirement, so I didn't bother working on moving that into the core engine.

**Comment by RalfJung** [comment]
Chocolatey was down. @bors retry

**Pull request merged**